File: frontend/src/types/candidate.ts
export type CandidateStatus = "discovered" | "reviewed" | "promoted" | "rejected";

export interface CandidateSummary {
  id: number;
  ip_address: string;
  port: number;
  unit_id: number;
  status: CandidateStatus;
  device_type_guess: string | null;
  confidence_score: number | null;
  vendor_guess: string | null;
  updated_at: string;
}

export interface CandidateProbeResult {
  id: number;
  candidate_id: number;
  register_address: number;
  function_code: string;
  data_type: string;
  raw_value: string | null;
  decoded_value: number | null;
  valid: boolean;
  created_at: string;
}

export interface CandidateDetails extends CandidateSummary {
  scan_result_id: number;
  notes: string | null;
  created_at: string;
  probe_results: CandidateProbeResult[];
}

export interface PromoteCandidatePayload {
  device_name: string;
  location?: string | null;
  description?: string | null;
}

export interface PromoteCandidateResponse {
  candidate_id: number;
  device_id: number;
  status: string;
  registers_created: number;
}

File: backend/app/schemas/candidate.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class CandidateSummary(BaseModel):
    id: int
    ip_address: str
    port: int
    unit_id: int
    status: str
    device_type_guess: Optional[str]
    confidence_score: Optional[float]
    vendor_guess: Optional[str]
    updated_at: str


class CandidateProbeResultRead(BaseModel):
    id: int
    candidate_id: int
    register_address: int
    function_code: str
    data_type: str
    raw_value: Optional[str]
    decoded_value: Optional[float]
    valid: bool
    created_at: str


class CandidateDetails(CandidateSummary):
    scan_result_id: int
    notes: Optional[str]
    created_at: str
    probe_results: list[CandidateProbeResultRead]


class CandidateProbeResponse(BaseModel):
    candidate_id: int
    probe_status: str
    valid_registers_found: int
    device_type_guess: str
    confidence_score: float


class CandidatePromoteRequest(BaseModel):
    device_name: str = Field(min_length=1, max_length=160)
    location: Optional[str] = None
    description: Optional[str] = None


class CandidatePromoteResponse(BaseModel):
    candidate_id: int
    device_id: int
    status: str
    registers_created: int


class CandidateRejectResponse(BaseModel):
    candidate_id: int
    status: str
    message: str
